semstr/conversion/ptb.py

```diff
@@ -0,0 +1,282 @@
+from ucca import core, layer0, layer1
+from ucca.layer1 import EdgeTags
+from .format import FormatConverter
+
+import sys
+import re
+
+
+class PtbConverter(FormatConverter):
+    def __init__(self):
+        self.node_by_id = {}
+        self.terminals = []
+        self.pending_nodes = []
+        self.node_ids_with_children = set()
+        self.sentence_id = 1
+        self.passage_id = 3
+
+    def _build_passage(self, stream):
+        # p = core.Passage(self.sentence_id or self.passage_id)
+        p = core.Passage(self.passage_id)
+        l0 = layer0.Layer0(p)
+        l1 = layer1.Layer1(p)
+        paragraph = 1
+
+        next(self.parse(stream))
+
+        # add normal nodes
+        self.pending_nodes = list(reversed(self.pending_nodes))
+        while self.pending_nodes:
+            for i in reversed(range(len(self.pending_nodes))):
+                parent_id, edge_tag, node_id = self.pending_nodes[i]
+                parent = self.node_by_id.get(parent_id, -1)
+                if parent != -1:
+                    del self.pending_nodes[i]
+                    implicit = node_id not in self.node_ids_with_children
+                    node = l1.add_fnode(parent, edge_tag, implicit=implicit)
+                    if edge_tag == EdgeTags.Punctuation:
+                        node.tag = layer1.NodeTags.Punctuation
+                    self.node_by_id[node_id] = node
+
+        # add terminals
+        for text, tag, edge_tag, parent_id in self.terminals:
+            punctuation = (tag == layer0.NodeTags.Punct)
+            terminal = l0.add_terminal(text=text, punct=punctuation, paragraph=paragraph)
+            try:
+                parent = self.node_by_id[parent_id]
+            except KeyError as e:
+                raise ValueError("Terminal ('%s') with bad parent (%s) in passage %s" % (text, parent_id, p.ID)) from e
+            if parent is None:
+                print("Terminal is a child of the root: '%s'" % text, file=sys.stderr)
+                parent = l1.add_fnode(parent, edge_tag)
+            if edge_tag != EdgeTags.Terminal:
+                print("Terminal with incoming %s edge: '%s'" % (edge_tag, text), file=sys.stderr)
+            parent.add(EdgeTags.Terminal, terminal)
+        return p
+
+    def from_format(self, stream, passage_id, return_original=False, **kwargs):
+        self.passage_id = passage_id
+        passage = self._build_passage(stream)
+        passage.extra["format"] = kwargs.get("format") or "ptb"
+        # yield (passage, self.lines_read, passage.ID) if return_original else passage
+        yield passage
+        self.node_by_id = None
+        self.lines_read = []
+
+    def parse(self, line_or_lines):
+        def istok(t, i):
+            return getattr(t, 'token_id', None) is i
+
+        stack = []
+        node_id = 0
+        for tok in lex(line_or_lines):
+            if tok.token_id is LPAREN_TOKEN:
```

Comment:
The content of these two `if` is the same, so better join them as `if tok.token_id is LPAREN_TOKEN or tok.token_id is STRING_TOKEN`.
Also, why `is` here and not `==`?